aspnetcore/test/middleware.md

```diff
@@ -0,0 +1,79 @@
+---
+title: Test ASP.NET Core middleware
+author: cross
+description: Learn how to test ASP.NET Core middleware with TestServer.
+ms.author: riande
+ms.custom: mvc
+ms.date: 4/05/2019
+no-loc: [Blazor, "Identity", "Let's Encrypt", Razor, SignalR]
+uid: test/middleware
+---
+# Test ASP.NET Core middleware
+
+By [Chris Ross](https://github.com/Tratcher)
+
+Middleware can be tested in isolation using the <xref:Microsoft.AspNetCore.TestHost.TestServer>. Instantiate an app pipeline containing only the components that you need to test and send custom requests to verify middleware behavior.
+
+The requests:
+
+* Are sent in-memory rather than being serialized over the network.
+* Avoid additional concerns like port management and HTTPS certificates.
+
+Exceptions in the middleware can flow directly back to the calling test.
+
+Set up your test app, add your own middleware and services, and send a request using <xref:System.Net.Http.HttpClient>:
+
+```csharp
+[Fact]
+public async Task GenericCreateAndStartHost_GetTestServer()
+{
+    using var host = await new HostBuilder()
+        .ConfigureWebHost(webBuilder =>
+        {
+            webBuilder
+                .UseTestServer()
+                .Configure(app => { });
+        })
+        .StartAsync();
+
+    var response = await host.GetTestServer().CreateClient().GetAsync("/");
+
+    Assert.Equal(HttpStatusCode.NotFound, response.StatusCode);
+}
+```
+
+A test app can also send a request using [SendAsync(Action\<HttpContext>, CancellationToken)](xref:Microsoft.AspNetCore.TestHost.TestServer.SendAsync%2A):
+
+```csharp
+[Fact]
+public async Task ExpectedValuesAreAvailable()
+{
+    var builder = new WebHostBuilder().Configure(app => { });
```

Comment:
Don't use any WebHostBuilder examples, only HostBuilder. The SendAsync example picks up after the HostBuilder sample like this:
```
var server = host.GetTestServer();
server.BaseAddress = new Uri("https://example.com/A/Path/");
var context = await server.SendAsync(c =>
    {
        c.Request.Method = HttpMethods.Post;
        c.Request.Path = "/and/file.txt";
        c.Request.QueryString = new QueryString("?and=query");
    });
...
```